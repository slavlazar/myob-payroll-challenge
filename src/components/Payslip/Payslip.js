import React from 'react';
import PropTypes from 'prop-types';

import Header from './Header';
import Generator from './Generator';
import Generated from './Generated';
import Footer from './Footer';

import { Container, Heading } from '../shared/ui';
import Wrapper, { MainBody, BodyWrapper } from './Payslip.style';

/**
 * Payslip component rendering our payslip view with conditional rendering
 * of generator form (payslip create form) or generated screen
 * @param {object} Object containing payslip data along with submit and
 * reset actions from our Payslip store
 */
const Payslip = ({ payslip, handleSubmit, handleReset }) => {
  const { loading, error, result } = payslip;

  return (
    <Wrapper>
      <Header />
      <MainBody id="main">
        <BodyWrapper>
          <Container>
            <Heading>Pay Slip Generator</Heading>{' '}
            {result === null || !result ? (
              <Generator
                handleSubmit={handleSubmit}
                loading={loading}
                error={error}
              />
            ) : (
              <Generated payslip={payslip} handleReset={handleReset} />
            )}
          </Container>
        </BodyWrapper>
      </MainBody>
      <Footer />
    </Wrapper>
  );
};

Payslip.propTypes = {
  /** Payslip object with employee payslip data generated */
  payslip: PropTypes.object.isRequired,
  /** Function to handle submit of payslip */
  handleSubmit: PropTypes.func.isRequired,
  /** Function to handle reset of generate payslip to generate new one */
  handleReset: PropTypes.func.isRequired,
};

export default Payslip;
