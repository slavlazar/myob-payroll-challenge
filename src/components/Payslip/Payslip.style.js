import styled from 'styled-components';

const PayslipWrapper = styled.div`
  display: flex;
  min-height: 100vh;
  flex-direction: column;
`;

export const MainBody = styled.main`
  display: flex;
  flex: 1;
`;

export const BodyWrapper = styled.div`
  flex: 1;
  padding: 5rem 0;
`;

export default PayslipWrapper;
