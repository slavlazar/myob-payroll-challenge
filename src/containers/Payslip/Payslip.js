import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import actions from '../../actions/payslip';
import Payslip from '../../components/Payslip';
import {
  payslipSelector,
  payslipResultSelector,
} from '../../selectors/payslip';

const { doGeneratePayslip, doResetPayslip } = actions;

/**
 * Payslip container to handle connection to Redux store
 * and handle Payslip component actions
 * @param {object} Params from Redux store of Payslip
 */
const PayslipContainer = ({
  Payslip: PayslipState,
  dispatchDoGeneratePayslip,
  dispatchDoResetPayslip,
}) => {
  const handlePayslipGenerate = employeeDetails => {
    dispatchDoGeneratePayslip(employeeDetails);
  };

  const handleResetPayslip = () => {
    dispatchDoResetPayslip();
  };

  return (
    <Payslip
      payslip={PayslipState}
      handleSubmit={handlePayslipGenerate}
      handleReset={handleResetPayslip}
    />
  );
};

PayslipContainer.propTypes = {
  /** Payslip object */
  Payslip: PropTypes.object.isRequired,
  /** Function to generate payslip from store action creators */
  dispatchDoGeneratePayslip: PropTypes.func.isRequired,
  /** Function to reset payslip from store action creators */
  dispatchDoResetPayslip: PropTypes.func.isRequired,
};

const mapStateToProps = state => ({
  Payslip: {
    ...payslipSelector(state),
    ...payslipResultSelector(state),
  },
});

export default connect(
  mapStateToProps,
  {
    dispatchDoGeneratePayslip: doGeneratePayslip,
    dispatchDoResetPayslip: doResetPayslip,
  }
)(PayslipContainer);
