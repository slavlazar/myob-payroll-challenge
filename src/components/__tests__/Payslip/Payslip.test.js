import React from 'react';
import { shallow } from 'enzyme';

import Payslip from '../../Payslip/Payslip';

const shallowWrapper = (props = {}) => {
  const allProps = {
    payslip: {},
    handleSubmit: () => {},
    handleReset: () => {},
    ...props,
  };

  return shallow(<Payslip {...allProps} />);
};

describe('Components: Payslip', () => {
  it('renders correctly', () => {
    const wrapped = shallowWrapper();
    expect(wrapped).toMatchSnapshot();
  });
});
