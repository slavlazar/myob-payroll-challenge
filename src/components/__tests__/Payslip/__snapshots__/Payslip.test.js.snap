// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components: Payslip renders correctly 1`] = `
<styled.div>
  <Header />
  <styled.main
    id="main"
  >
    <styled.div>
      <styled.div>
        <Heading
          noMargin={false}
          textAlign="left"
          type="h1"
        >
          Pay Slip Generator
        </Heading>
        \\
         
        <Generator
          error={null}
          handleSubmit={[Function]}
          loading={false}
        />
      </styled.div>
    </styled.div>
  </styled.main>
  <Footer />
</styled.div>
`;
